### PR TITLE
Fix address pool syncing and add new RPC commands for the address pools

### PR DESCRIPTION
**Pull request by cjepson**
There were major bugs affecting the synchronization of the address
pools on start up relating to the bad variables or legacy code. Much
of the address pool code has been rewritten along with the syncing
code to simplify the syncing of the address pool index to the last
used address.

New RPC commands have been added to help simplify and clarify address
pool index synchronization. These include accountaddressindex,
accountfetchaddresses, and accountsyncaddressindex.
getaddressesbyaccount has been corrected so that it now returns only
up to the addresses used by the address pool, fixing #113.

A new wallet RPC to check the global state of the wallet was added
called walletinfo.

### CODE REVIEW

**Comment by jcvernaleo** [comment]
Works in my testing but still needs a squash.
tACK

**Comment by cjepson** [comment]
Updated to add a missed check for the address pool existing that was missing from the 0.0.9 release.

**Comment by jcvernaleo** [comment]
The extra check doesn't seem to break anything for me so still good other than the squash.

**Comment by cjepson** [comment]
Accounts couldn't be created because waddrmgr has no address buffers for them, fixed.

File: wallet/addresspool.go [hunk 1]
```diff
@@ -295,6 +338,75 @@ func (w *Wallet) CheckAddressPoolsInitialized(account uint32) error {
...
+	index uint32) error {
+	// Sanity checks.
+	err := w.CheckAddressPoolsInitialized(account)
+	if err != nil {
+		return err
+	}
+	var addrPool *addressPool
+	switch branch {
+	case waddrmgr.ExternalBranch:
+		addrPool = w.addrPools[account].external
```
Comment (alexlyp) [inline wallet/addresspool.go]: why is this done outside of the mutex locks?

File: wallet/sync.go [hunk 2]
```diff
@@ -375,63 +375,38 @@ func (w *Wallet) scanAddressIndex(start int, end int, account uint32,
...
-		counter++
 	}
 
-	if isSynced {
-		return nil
-	}
-
-	return fmt.Errorf("failed to sync to address %v during address rescan",
-		addr.String())
+	// We can't find any used addresses. The wallet is
```
Comment (alexlyp) [inline wallet/sync.go]: Should be account and branch from args is unused?  doesn't mean entire wallet is unused

**Comment by alexlyp** [comment]
tACK
after review everything looks good.  just a couple questions about a few locks and an incorrect comment.
once those are resolved, good to merge

**Pull request merged**